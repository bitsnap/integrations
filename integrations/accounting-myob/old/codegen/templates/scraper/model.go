package scraper

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
	. "github.com/bitsnap/bitsnap/packages/go-util"
	"github.com/gocolly/colly/v2"
)

type Model struct {
	Name       string
	ParentName string
	Attributes []Attribute
}

func (m *Model) String() string {
	attrs := make([]string, 0, len(m.Attributes))
	for _, attr := range m.Attributes {
		reqSuffix := "`"
		if attr.Required {
			reqSuffix = " validation:\"required\"`"
		}
		attrs = append(attrs, "\t"+attr.Name+"\t"+attr.Type+" `json:\""+attr.Name+"\""+reqSuffix+"\t // "+attr.Description)
	}

	return "type " + m.Name + " struct {\n" + strings.Join(attrs, "\n") + "\n}"
}

var knownModels = make(map[string]*Model)

var lengthRegex = regexp.MustCompile(`.*\(\s*(\d+)\s*\)\s*$`)

func ScrapeModel(modelName, parentName, page string) (*Model, error) {
	collector := colly.NewCollector(
		colly.AllowedDomains("developer.myob.com", "myob.com"),
		colly.CacheDir("./.cache"),
	)

	attributes := make([]Attribute, 0, 10)

	prefix := modelName
	if parentName != "" {
		prefix = parentName + Capitalize(modelName)
	}

	if knownModels[Capitalize(modelName)] != nil {
		return knownModels[Capitalize(modelName)], nil
	}

	collector.OnHTML("div.documentation", func(docs *colly.HTMLElement) {
		docs.DOM.Find("h4").FilterFunction(func(_ int, header *goquery.Selection) bool {
			return strings.TrimSpace(strings.ToLower(header.Text())) == "attribute details"
		}).First().NextAllFiltered("ul.apiDocs").First().Each(func(_ int, table *goquery.Selection) {
			table.Find("li.attribute").Each(func(_ int, attr *goquery.Selection) {
				if cls, ok := attr.Parent().Parent().Attr("class"); ok && !strings.Contains(cls, "subgroupDetails") {
					t := strings.TrimSpace(attr.Find("span.type").First().Text())
					name := strings.TrimSpace(strings.Replace(attr.Text(), t, "", 1))
					t = strings.ToLower(t)

					desc := strings.TrimSpace(attr.NextAllFiltered("li.description").First().Text())
					length := 0
					if lengthRegex.MatchString(t) {
						l := lengthRegex.FindStringSubmatch(t)
						length, _ = strconv.Atoi(l[len(l)-1])

						if length > 0 {
							t = strings.TrimSpace(strings.Replace(t, "("+strconv.Itoa(length)+")", "", -1))
						}
					}

					if t == "" {
						t = "models." + Capitalize(prefix)
					} else {
						t = formatType(t)
					}

					attributes = append(attributes, *NewAttribute(name, t, desc, attr.HasClass("required"), length))
				}
			})
		})
	})

	collector.OnError(func(response *colly.Response, err error) {
		Logger().Errorw(response.Request.URL.String(), err)
	})

	err := collector.Visit(page)
	if err != nil {
		return nil, err
	}

	m := &Model{
		Name:       Capitalize(modelName),
		ParentName: Capitalize(parentName),
		Attributes: attributes,
	}

	if _, ok := knownModels[m.Name]; ok {
		Logger().Infow("already scraped model", "model", m)
	} else {
		knownModels[m.Name] = m
	}

	return m, nil
}
