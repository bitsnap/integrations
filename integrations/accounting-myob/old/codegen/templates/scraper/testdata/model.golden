# Generated by goldga. DO NOT EDIT.
[snapshots]
"Model attributes scraping when traversing documentation pages should be able to scrape model from documentation page" = '''
(string) (len=3794) "type CustomerContract struct {\n\tUID\tstring `json:\"UID\" validation:\"required\"`\t // Unique customer contact identifier in the form of a guid.ONLY required for updating an existing contact NOT required when creating a new contact.\n\tCompanyName\tstring `json:\"CompanyName\" validation:\"required\"`\t // Company name of the customer contact. ONLY required when creating a company contact with IsIndividual set to false.\n\tLastName\tstring `json:\"LastName\" validation:\"required\"`\t // Last name for an individual contact.\n\tFirstName\tstring `json:\"FirstName\" validation:\"required\"`\t // First name for an individual contact.\n\tIsIndividual\tboolean `json:\"IsIndividual\" validation:\"required\"`\t // True indicates the customer contact represents an individual.False indicates the customer contact represents a company.\n\tDisplayID\tstring `json:\"DisplayID\"`\t // Customer contact card id. { 'DisplayID': 'CUS000001' }\n\tIsActive\tboolean `json:\"IsActive\" validation:\"required\"`\t // True indicates the customer contact is active.False indicates the customer contact is inactive.\n\tAddress\tmodels.ContactCustomerContract `json:\"Address\"`\t // The following set of information pulls through details for a contact address\n\tNotes\tstring `json:\"Notes\"`\t // Notes for the customer contact.\n\tIdentifier\tmodels.ContactCustomerContract `json:\"Identifier\"`\t // This is an AccountRight only field.The following set of information pulls through details for a contact identifier\n\tCustomList1\tmodels.ContactCustomerContract `json:\"CustomList1\"`\t // This is an AccountRight only field. The following set of information pulls through details for a contact custom list\n\tCustomList2\tmodels.ContactCustomerContract `json:\"CustomList2\"`\t // This is an AccountRight only field .The following set of information pulls through details for a contact custom list\n\tCustomList3\tmodels.ContactCustomerContract `json:\"CustomList3\"`\t // This is an AccountRight only field. The following set of information pulls through details for a contact custom list\n\tCustomField1\tmodels.ContactCustomerContract `json:\"CustomField1\"`\t // This is an AccountRight only field. The following set of information pulls through details for a contact custom field\n\tCustomField2\tmodels.ContactCustomerContract `json:\"CustomField2\"`\t // This is an AccountRight only field. The following set of information pulls through details for a contact custom field\n\tCustomField3\tmodels.ContactCustomerContract `json:\"CustomField3\"`\t // This is an AccountRight only field. The following set of information pulls through details for a contact custom field\n\tCurrentBalance\tdecimal (13.6) `json:\"CurrentBalance\"`\t // Customer contact balance.\n\tSellingDetails\tmodels.ContactCustomerContract `json:\"SellingDetails\" validation:\"required\"`\t // The following set of information pulls through default customer selling details.\n\tPaymentDetails\tmodels.ContactCustomerContract `json:\"PaymentDetails\"`\t // The following set of information pulls through customer payment details.\n\tForeignCurrency\tmodels.ContactCustomerContract `json:\"ForeignCurrency\"`\t // This is an AccountRight only field.Null if no foreign currency is set.\n\tLastModified\t*time.Time `json:\"LastModified\"`\t // Extracts the last modification date for the customer contact resource.\n\tPhotoURI\tstring `json:\"PhotoURI\"`\t // This is an AccountRight only field. Uniform resource identifier associated with a photo image.\n\tURI\tstring `json:\"URI\"`\t // Uniform resource identifier encompasses all types of names and addresses that refer to objects on the web.\n\tRowVersion\tstring `json:\"RowVersion\" validation:\"required\"`\t // Number value that changes upon a record update, can be used for change control but does does not preserve a date or a time. ONLY required for updating an existing contact NOT required when creating a new contact.\n}"
'''
